### PR TITLE
fix: check for installed serve-d version + improve get.serve-d 

### LINKED ISSUE #38: Install serve-d on demand
Integrate [`get-serve-d`](https://github.com/Pure-D/atom-ide-d/blob/main/scripts/get-serve-d.js) script into the the main code of the package instead of being an offline script

### PR DESCRIPTION
**Pull request by aminya**

### CODE REVIEW

File: src/get-serve-d.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    "Pure-D",
+    "serve-d",
+    distFolder,
+    undefined,
+    undefined,
+    true
+  )) as unknown) as string[]
+  for (const asset of assets) {
+    const platform = basename(asset).match(/windows|linux|osx/)[0]
+    extname(asset)
```
Comment (WebFreak001) [inline src/get-serve-d.ts:22]: useless statement?

File: src/get-serve-d.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  await ensureDir(distFolder)
+  const assets = ((await downloadRelease(
+    "Pure-D",
+    "serve-d",
+    distFolder,
+    undefined,
+    undefined,
+    true
+  )) as unknown) as string[]
+  for (const asset of assets) {
```
Comment (WebFreak001) [inline src/get-serve-d.ts:20]: should it really download all assets for a given release?

File: src/get-serve-d.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import downloadRelease from "@terascope/fetch-github-release"
+import { join, dirname, basename, extname } from "path"
+import { remove, ensureDir } from "fs-extra"
+import decompress from "decompress"
+import decompressTarxz from "decompress-tarxz"
+
+// function to download serve-d binaries from GitHub
+export async function getServeD() {
```
Comment (WebFreak001) [inline src/get-serve-d.ts:8]: I think nightly & pre-release support would match the vscode extension well

File: src/get-serve-d.ts (hunk 2 above)
Comment (aminya) [reply src/get-serve-d.ts:20]: No, I am going to fix this.

Comment (aminya) [reply src/get-serve-d.ts]: Fixed

File: src/get-serve-d.ts (hunk 1 above)
Comment (aminya) [reply src/get-serve-d.ts]: Fixed

File: src/get-serve-d.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  if (process.platform !== "win32") {
+    decompressPlugins.push(await import("decompress-tarxz"))
+  }
+
+  const assets = ((await downloadRelease(
+    /* username */ "Pure-D",
+    /* repo */ "serve-d",
+    /* download folder */ distFolder,
+    /* filter release */ undefined,
+    /* filter asset */ (asset) => asset.name.indexOf(platform) >= 0,
```
Comment (WebFreak001) [inline src/get-serve-d.ts:26]: this test should probably be a little more exact

you might want to copy the function from https://github.com/Pure-D/code-d/blob/master/src/installer.ts#L261

**Review by aminya — COMMENTED** [review]
Windows installation fails because tarxz runs node-gyp and this requires Visual Studio Installation. 

I think it would be easier to bundle serve-d with the package

**Comment by WebFreak001** [comment]
on windows the releases are zips, there is no need for tarxz.

In code-d I use AdmZip to unzip these releases. For .tar.xz I invoke `tar` as command (assuming posix compliant system)

**Comment by aminya** [comment]
"@terascope/fetch-github-release" itself supports extracting zip. The decompress plugin is only needed for tarxz. I wish the assets were consistent, and they all used zip.

**Comment by WebFreak001** [comment]
the releases are currently more or less following what is usually used on each platform, I believe that makes the most sense for end users manually downloading releases.

The code in code-d for extracting tar.xz & tar.gz is fairly simple, would it be a solution to use that code instead of the decompress plugin? https://github.com/Pure-D/code-d/blob/master/src/installer.ts#L441-L462

File: src/get-serve-d.ts (hunk 4 above)
Comment (aminya) [reply src/get-serve-d.ts:28]: But do you build for other architectures? If you don't there is no need to add code for it.

**Comment by aminya** [comment]
How do you check if serve-d is up to date?

**Comment by aminya** [comment]
I feel like this approach is just more complex for no reason (e.g. all these manual downloading, decompressing, checking for versions, etc). 

The better solution is to make an npm package for serve-d, so it can be easily installed as a dependency. 

For now, I am going to bundle serve-d exe as part of the package.

File: src/get-serve-d.ts (hunk 4 above)
Comment (WebFreak001) [reply src/get-serve-d.ts]: I don't build for other architectures yet, but vscode is supported on ARM, so if there is an ARM user (e.g. coders on cheap linux ARM laptops or raspberry pi) they don't download a release that is unsupported anyway.

It probably makes sense to move all the release downloading & manual compilation into an NPM package which can be used by all JS based extensions.

**Pull request merged**